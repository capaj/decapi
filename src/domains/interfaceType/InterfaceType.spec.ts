import {
  Field,
  Query,
  ObjectType,
  compileSchema,
  SchemaRoot,
  InterfaceType,
} from '..'
import { GraphQLInt, graphql, printSchema } from 'graphql'

describe('InterfaceType', () => {
  @InterfaceType()
  abstract class Vehicle {
    @Field({ type: GraphQLInt })
    windows: number

    @Field({ type: GraphQLInt })
    seats: number
  }

  @ObjectType({ implements: Vehicle })
  class Car implements Vehicle {
    windows: number
    seats: number
    @Field()
    wheels: number
  }

  @ObjectType({ implements: Vehicle })
  class Boat implements Vehicle {
    windows: number
    seats: number

    @Field()
    propellers: number
  }

  @SchemaRoot()
  class FooSchema {
    @Query({ type: [Vehicle] })
    vehicles(): Vehicle[] {
      const car = new Car()
      car.seats = 4
      car.windows = 6
      car.wheels = 4

      const boat = new Boat()
      boat.seats = 150
      boat.windows = 30
      return [car]
    }
    @Query({ type: Boat })
    boat(): Boat {
      return null
    }
  }

  const schema = compileSchema(FooSchema)

  it('should expose GQL interface', async () => {
    expect(printSchema(schema)).toMatchSnapshot()
  })
  it('should resolve', async () => {
    await graphql(
      schema,
      `
        {
          vehicles {
            seats
            wheels
          }
        }
      `,
    )
  })
})
