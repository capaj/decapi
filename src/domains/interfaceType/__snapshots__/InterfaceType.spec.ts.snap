// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InterfaceType should expose GQL interface type 2`] = `
"\\"\\"\\"a vehicle interface for a basic spec\\"\\"\\"
interface IVehicle {
  windows: Int
  seats: Int
}

\\"\\"\\"a vehicle interface for a basic spec\\"\\"\\"
interface IEntity {
  id: Int
}

type Car implements IVehicle & IEntity {
  id: Int
  wheels: Float
  windows: Int
  seats: Int
}

type Boat implements IVehicle & IEntity {
  id: Int
  propellers: Float
  windows: Int
  seats: Int
}

type Katamaran implements IVehicle & IEntity {
  id: Int
  propellers: Float
  hulls: Float
  windows: Int
  seats: Int
}

type Query {
  vehicles: [IVehicle!]
  boat: Boat
  car: Car
}"
`;

exports[`InterfaceType should resolve correctly 1`] = `
Object {
  "data": Object {
    "vehicles": Array [
      Object {
        "__typename": "Car",
        "seats": 4,
        "wheels": 4,
        "windows": 6,
      },
      Object {
        "__typename": "Boat",
        "propellers": null,
        "seats": 150,
        "windows": 30,
      },
    ],
  },
}
`;

exports[`InterfaceType should resolve correctly even when extending another implemetor type 1`] = `
Object {
  "data": Object {
    "boat": Object {
      "__typename": "Boat",
      "propellers": null,
      "seats": 150,
      "windows": 30,
    },
    "car": Object {
      "__typename": "Car",
      "seats": 4,
      "wheels": 4,
      "windows": 6,
    },
    "vehicles": Array [
      Object {
        "__typename": "Car",
        "seats": 4,
        "wheels": 4,
        "windows": 6,
      },
      Object {
        "__typename": "Boat",
        "propellers": null,
        "seats": 150,
        "windows": 30,
      },
      Object {
        "__typename": "Katamaran",
        "hulls": 2,
        "seats": null,
        "windows": null,
      },
    ],
  },
}
`;
